lib/opam_solve.ml

```diff
@@ -251,6 +254,84 @@ module Make_solver (Context : OPAM_MONOREPO_CONTEXT) :
       rolemap []
     |> List.filter_map ~f:Solver.package_name
 
+  let determine_version_restriction component =
+    let open Option.O in
+    let restrictions =
+      component |> Solver.Diagnostics.Component.notes
+      |> List.map ~f:(function
+           | Solver.Diagnostics.Note.UserRequested restriction ->
+               [ restriction ]
+           | Restricts (_other_role, _impl, restrictions) -> restrictions
```

Comment:
At some point I think we should collect `other_role` here as it would help improving the error message!